Add PUT Repository High Level REST API

hub-cap opened the pull request:
This commit adds PUT Repository, the associated docs and tests for the
high level REST API client. A few small changes to the PutRepository
Request and Response went into the commit as well.

Relates #27205

elasticmachine commented:
Pinging @elastic/es-core-infra

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java:
```diff
@@ -656,6 +657,18 @@ static Request getRepositories(GetRepositoriesRequest getRepositoriesRequest) {
+        parameters.withMasterTimeout(putRepositoryRequest.masterNodeTimeout());
+        parameters.withTimeout(putRepositoryRequest.timeout());
+
+        request.setEntity(createEntity(putRepositoryRequest, REQUEST_BODY_CONTENT_TYPE));
```
I see in the corresponding REST action that we also support the `verify` parameter. I guess we should support it here too?

hub-cap replied:
yea definitely, ty for pointing it out. Cant believe i missed it!

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java:
```diff
@@ -67,4 +67,28 @@ public void getRepositoriesAsync(GetRepositoriesRequest getRepositoriesRequest,
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-snapshots.html"> Snapshot and Restore
+     * API on elastic.co</a>
+     */
+
```
remove this empty line?

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java:
```diff
@@ -1448,6 +1451,18 @@ public void testGetRepositories() {
+        String repository = "repo";
+        PutRepositoryRequest putRepositoryRequest = new PutRepositoryRequest(repository);
+        putRepositoryRequest.type(FsRepository.TYPE);
+        putRepositoryRequest.settings("{\"location\": \".\"}", XContentType.JSON);
```
would it be possible to randomize the values that the request holds like in other tests?

javanna replied:
I assumed that there is no problem setting values and checking that the output of the conversion from high-level request to low-level request is the expected one. We don't validate etc. I would do only what is straight-forward.

javanna commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/put/PutRepositoryResponse.java:
```diff
@@ -49,4 +58,8 @@ public void writeTo(StreamOutput out) throws IOException {
 
+    public static PutRepositoryResponse fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
```
can you add a test for this class although very simple?

javanna commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/put/PutRepositoryRequest.java:
```diff
@@ -232,4 +233,17 @@ public void writeTo(StreamOutput out) throws IOException {
+        builder.endObject();
+
+        builder.endObject();
+        return builder;
```
can you add a unit test for this method?

nik9000 commented on client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java:
```diff
@@ -67,4 +67,28 @@ public void getRepositoriesAsync(GetRepositoriesRequest getRepositoriesRequest,
+     * API on elastic.co</a>
+     */
+    public void putRepositoryAsync(PutRepositoryRequest putRepositoryRequest,
+                                   ActionListener<PutRepositoryResponse> listener, Header... headers){
```
Add a space after `)` and before `{` please!

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java:
```diff
@@ -67,4 +67,28 @@ public void getRepositoriesAsync(GetRepositoriesRequest getRepositoriesRequest,
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-snapshots.html"> Snapshot and Restore
+     * API on elastic.co</a>
+     */
+    public void putRepositoryAsync(PutRepositoryRequest putRepositoryRequest,
```
in the spec we call this "snapshot.create_repository" . Could you rename the methods to createRepository please?

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java:
```diff
@@ -1448,6 +1451,18 @@ public void testGetRepositories() {
         assertThat(expectedParams, equalTo(request.getParameters()));
     }
 
+    public void testPutRepository() throws IOException {
```
testCreateRepository?

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SnapshotClientDocumentationIT.java:
```diff
@@ -58,7 +62,114 @@
-    private static final String testRepository = "test_repository";
+    private static final String repositoryName = "test_repository";
+
+    public void testSnapshotPutRepository() throws IOException {
```
testSnapshotCreateRepository?

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SnapshotClientDocumentationIT.java:
```diff
@@ -58,7 +62,114 @@
+        assertTrue(acknowledged);
+    }
+
+    public void testSnapshotPutRepositoryAsync() throws InterruptedException {
```
testSnapshotCreateRepositoryAsync

javanna commented on docs/java-rest/high-level/snapshot/put_repository.asciidoc:
```diff
@@ -0,0 +1,139 @@
+[[java-rest-high-snapshot-put-repository]]
+=== Snapshot Put Repository API
```
in the docs as well, can you rename put repository to create repository. We leave requests and responses as they are as we reuse them from the transport client, but for the rest we should follow the spec naming like other language clients already do.

hub-cap replied:
I went ahead and changed it in all the headings and links (eg, `[[java-rest-high-snapshot-create-repository]]`)

hub-cap commented:
@javanna ive addressed comments, pls review. I have also added repository request/response tests so i can cover most of the new things im adding to these request/responses in this and subsequent PRs.

javanna commented on server/src/test/java/org/elasticsearch/action/admin/cluster/repositories/RepositoryResponseTests.java:
```diff
@@ -0,0 +1,38 @@
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class RepositoryResponseTests extends ESTestCase {
```
could this extend AbstractXContentTestCase or AbstractStreamableXContentTestCase ?

lcawl commented:
I think the docs/java-rest/high-level/snapshot/create_repository.asciidoc is missing from the 6.x branch. Is that intentional?  It's causing some documentation build issues.

hub-cap commented:
@lcawl I had not yet pushed it. I was in backport hell for a while today. I pushed it just now.
